R/layers.R

```diff
@@ -237,12 +241,124 @@ addMarkers = function(
   options = markerOptions(),
   data = getMapData(map)
 ) {
-  options$icon = icon
+  options$icon = L.icon(evalFormula(icon, data))
+  options$iconGroup = attr(options$icon, 'iconGroup', exact = TRUE)
   pts = derivePoints(data, lng, lat, missing(lng), missing(lat), "addMarkers")
   invokeMethod(map, data, 'addMarkers', pts$lat, pts$lng, layerId, options, popup) %>%
     expandLimits(pts$lat, pts$lng)
 }
 
+L.icon = function(options, embed = TRUE) {
+  if (!is.list(options)) return(options)
+  if (is.null(names(options))) {
+    if (length(options) == 0) return()
+    # in theory we need to check all sub-lists and all must be named
+    if (is.null(names(options[[1]])))
+      stop('The data for an individual icon must be a named list')
+    # it is already a list of icons, so no further base64 encoding
+    return(options)
+  }
+  image_uri = getFromNamespace('.b64EncodeFile', 'markdown')
```

Comment:
Ummm... I didn't realize you meant a *private* function from markdown... :)

Come to think of it, I might be bringing in Rcpp and a slew of native code for raster support, maybe we just copy in a base64 encoding routine while we're at it. Or is there a reason you didn't go with base64enc?